### DISCUSSION TITLE
Generating PDF's out of Editor

### DISCUSSION THREAD

**Discussion by PiochU19** [react-page/react-page#1224] [Q&A] [answered]
Hi, I'm wondering what's the best way to generate PDF from Editor. I want my react-page Editor to be PDF's generator.
I have seen discussions about it, but none of this worked for me. Frontend is in React of course, and backend in Python.

Solutions I have tried:
1. Generating PDF at the backend. For that I need html and css. I was able to get html, but unable to get css styles, so the PDF is not looking as expected.
2. [react-pdf](https://react-pdf.org/) + [react-pdf-html](https://github.com/danomatic/react-pdf-html). In this solution I was able to pass html (without images, there was some error don't remember exactly), but again there was no styling. I expected react-pdf to get styles directly from the page somehow, but it's not working like that.

Does anyone found a working solution for PDF integration? Thanks in advance.

**Comment by macrozone** [comment, marked as answer]
for https://github.com/danomatic/react-pdf-html you need to add a style-tag as described here: https://github.com/danomatic/react-pdf-html#rendering-react-components

for just rendering, not many stylings are needed (ignoring any custom component), its basically classes for cells and rows so that the column layout is working.

read-only output is done like this https://github.com/react-page/react-page/blob/master/packages/editor/src/renderer/HTMLRenderer.tsx

If you use custom cell components, it depends on what you use for styling there. if you use styled-components, you can actually collect the styles, see here https://styled-components.com/docs/advanced#example


If you want to go a full bottom-up approach, you could create some custom renderer and do something like the html renderer is doing https://github.com/react-page/react-page/blob/master/packages/editor/src/renderer/HTMLRenderer.tsx, but instead using react-pdf elements.

That means, that you have to re-declare all plugins you use (at least their id and Renderer) and have the Renderer return some react-pdf element. That of course does not make the pdf look exactly the same as the pdf, but that is probably not the goal anyway.. that is surely much more work, but you could define precisly how each cell should look like on a pdf.
